File: dart/test/scheme_test.dart
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:material_color_utilities/hct/hct.dart';
import 'package:material_color_utilities/scheme/scheme_content.dart';
import 'package:material_color_utilities/scheme/scheme_tonal_spot.dart';
import 'package:test/test.dart';

import './utils/color_matcher.dart';

void main() {
  test('blue light scheme', () {
    final scheme = SchemeTonalSpot(
      sourceColorHct: Hct.fromInt(0xff0000ff),
      isDark: false,
      contrastLevel: 0.0,
    );
    expect(scheme.primary, isColor(0xff555992));
  });

  test('blue dark scheme', () {
    final scheme = SchemeTonalSpot(
      sourceColorHct: Hct.fromInt(0xff0000ff),
      isDark: true,
      contrastLevel: 0.0,
    );
    expect(scheme.primary, isColor(0xffbec2ff));
  });

  test('3rd party light scheme', () async {
    final scheme = SchemeTonalSpot(
      sourceColorHct: Hct.fromInt(0xff6750a4),
      isDark: false,
      contrastLevel: 0.0,
    );

    expect(scheme.primary, isColor(0xff65558f));
    expect(scheme.secondary, isColor(0xff625b71));
    expect(scheme.tertiary, isColor(0xff7e5260));
    expect(scheme.surface, isColor(0xfffdf7ff));
    expect(scheme.onSurface, isColor(0xff1d1b20));
  });

  test('3rd party dark scheme', () async {
    final scheme = SchemeTonalSpot(
      sourceColorHct: Hct.fromInt(0xff6750a4),
      isDark: true,
      contrastLevel: 0.0,
    );
    expect(scheme.primary, isColor(0xffcfbdfe));
    expect(scheme.secondary, isColor(0xffcbc2db));
    expect(scheme.tertiary, isColor(0xffefb8c8));
    expect(scheme.surface, isColor(0xff141218));
    expect(scheme.onSurface, isColor(0xffe6e0e9));
  });

  test('light scheme from high chroma color', () async {
    final scheme = SchemeTonalSpot(
      sourceColorHct: Hct.fromInt(0xfffa2bec),
      isDark: false,
      contrastLevel: 0.0,
    );
    expect(scheme.primary, isColor(0xff814c77));
    expect(scheme.onPrimary, isColor(0xffffffff));
    expect(scheme.primaryContainer, isColor(0xffffd7f3));
    expect(scheme.onPrimaryContainer, isColor(0xff340830));
    expect(scheme.secondary, isColor(0xff6e5868));
    expect(scheme.onSecondary, isColor(0xffffffff));
    expect(scheme.secondaryContainer, isColor(0xfff8daee));
    expect(scheme.onSecondaryContainer, isColor(0xff271624));
    expect(scheme.tertiary, isColor(0xff815343));
    expect(scheme.onTertiary, isColor(0xffffffff));
    expect(scheme.tertiaryContainer, isColor(0xffffdbd0));
    expect(scheme.onTertiaryContainer, isColor(0xff321207));
    expect(scheme.error, isColor(0xffba1a1a));
    expect(scheme.onError, isColor(0xffffffff));
    expect(scheme.errorContainer, isColor(0xffffdad6));
    expect(scheme.onErrorContainer, isColor(0xff410002));
    expect(scheme.background, isColor(0xfffff7f9));
    expect(scheme.onBackground, isColor(0xff201a1e));
    expect(scheme.surface, isColor(0xfffff7f9));
    expect(scheme.onSurface, isColor(0xff201a1e));
    expect(scheme.surfaceVariant, isColor(0xffeedee7));
    expect(scheme.onSurfaceVariant, isColor(0xff4e444b));
    expect(scheme.outline, isColor(0xff80747b));
    expect(scheme.outlineVariant, isColor(0xffd2c2cb));
    expect(scheme.shadow, isColor(0xff000000));
    expect(scheme.scrim, isColor(0xff000000));
    expect(scheme.inverseSurface, isColor(0xff362e33));
    expect(scheme.inverseOnSurface, isColor(0xfffaedf3));
    expect(scheme.inversePrimary, isColor(0xfff3b2e4));
  });

  test('dark scheme from high chroma color', () async {
    final scheme = SchemeTonalSpot(
      sourceColorHct: Hct.fromInt(0xfffa2bec),
      isDark: true,
      contrastLevel: 0.0,
    );
    expect(scheme.primary, isColor(0xfff3b2e4));
    expect(scheme.onPrimary, isColor(0xff4d1f47));
    expect(scheme.primaryContainer, isColor(0xff67355e));
    expect(scheme.onPrimaryContainer, isColor(0xffffd7f3));
    expect(scheme.secondary, isColor(0xffdbbed1));
    expect(scheme.onSecondary, isColor(0xff3e2a39));
    expect(scheme.secondaryContainer, isColor(0xff554050));
    expect(scheme.onSecondaryContainer, isColor(0xfff8daee));
    expect(scheme.tertiary, isColor(0xfff5b9a5));
    expect(scheme.onTertiary, isColor(0xff4c2619));
    expect(scheme.tertiaryContainer, isColor(0xff663c2d));
    expect(scheme.onTertiaryContainer, isColor(0xffffdbd0));
    expect(scheme.error, isColor(0xffffb4ab));
    expect(scheme.onError, isColor(0xff690005));
    expect(scheme.errorContainer, isColor(0xff93000a));
    expect(scheme.onErrorContainer, isColor(0xffffdad6));
    expect(scheme.background, isColor(0xff181216));
    expect(scheme.onBackground, isColor(0xffecdfe5));
    expect(scheme.surface, isColor(0xff181216));
    expect(scheme.onSurface, isColor(0xffecdfe5));
    expect(scheme.surfaceVariant, isColor(0xff4e444b));
    expect(scheme.onSurfaceVariant, isColor(0xffd2c2cb));
    expect(scheme.outline, isColor(0xff9a8d95));
    expect(scheme.outlineVariant, isColor(0xff4e444b));
    expect(scheme.shadow, isColor(0xff000000));
    expect(scheme.scrim, isColor(0xff000000));
    expect(scheme.inverseSurface, isColor(0xffecdfe5));
    expect(scheme.inverseOnSurface, isColor(0xff362e33));
    expect(scheme.inversePrimary, isColor(0xff814c77));
  });

  test('light content scheme from high chroma color', () async {
    final scheme = SchemeContent(
      sourceColorHct: Hct.fromInt(0xfffa2bec),
      isDark: false,
      contrastLevel: 0.0,
    );

    expect(scheme.primary, isColor(0xffa7009e));
    expect(scheme.onPrimary, isColor(0xffffffff));
    expect(scheme.primaryContainer, isColor(0xffd100c6));
    expect(scheme.onPrimaryContainer, isColor(0xffffffff));
    expect(scheme.secondary, isColor(0xff9e2d93));
    expect(scheme.onSecondary, isColor(0xffffffff));
    expect(scheme.secondaryContainer, isColor(0xffff83ec));
    expect(scheme.onSecondaryContainer, isColor(0xff4a0045));
    expect(scheme.tertiary, isColor(0xffb51830));
    expect(scheme.onTertiary, isColor(0xffffffff));
    expect(scheme.tertiaryContainer, isColor(0xffd83546));
    expect(scheme.onTertiaryContainer, isColor(0xffffffff));
    expect(scheme.error, isColor(0xffba1a1a));
    expect(scheme.onError, isColor(0xffffffff));
    expect(scheme.errorContainer, isColor(0xffffdad6));
    expect(scheme.onErrorContainer, isColor(0xff410002));
    expect(scheme.background, isColor(0xfffff7f9));
    expect(scheme.onBackground, isColor(0xff251722));
    expect(scheme.surface, isColor(0xfffff7f9));
    expect(scheme.onSurface, isColor(0xff251722));
    expect(scheme.surfaceVariant, isColor(0xfff9daee));
    expect(scheme.onSurfaceVariant, isColor(0xff564050));
    expect(scheme.outline, isColor(0xff897081));
    expect(scheme.outlineVariant, isColor(0xffdcbed2));
    expect(scheme.shadow, isColor(0xff000000));
    expect(scheme.scrim, isColor(0xff000000));
    expect(scheme.inverseSurface, isColor(0xff3b2c37));
    expect(scheme.inverseOnSurface, isColor(0xffffebf6));
    expect(scheme.inversePrimary, isColor(0xffffabee));
  });

  test('dark content scheme from high chroma color', () async {
    final scheme = SchemeContent(
      sourceColorHct: Hct.fromInt(0xfffa2bec),
      isDark: true,
      contrastLevel: 0.0,
    );

    expect(scheme.primary, isColor(0xffffabee));
    expect(scheme.onPrimary, isColor(0xff5c0057));
    expect(scheme.primaryContainer, isColor(0xfffa2bec));
    expect(scheme.onPrimaryContainer, isColor(0xff000000));
    expect(scheme.secondary, isColor(0xffffabee));
    expect(scheme.onSecondary, isColor(0xff5c0057));
    expect(scheme.secondaryContainer, isColor(0xff840f7c));
    expect(scheme.onSecondaryContainer, isColor(0xffffd9f3));
    expect(scheme.tertiary, isColor(0xffffb3b2));
    expect(scheme.onTertiary, isColor(0xff680014));
    expect(scheme.tertiaryContainer, isColor(0xffff525f));
    expect(scheme.onTertiaryContainer, isColor(0xff000000));
    expect(scheme.error, isColor(0xffffb4ab));
    expect(scheme.onError, isColor(0xff690005));
    expect(scheme.errorContainer, isColor(0xff93000a));
    expect(scheme.onErrorContainer, isColor(0xffffdad6));
    expect(scheme.background, isColor(0xff1c0f19));
    expect(scheme.onBackground, isColor(0xfff4dceb));
    expect(scheme.surface, isColor(0xff1c0f19));
    expect(scheme.onSurface, isColor(0xfff4dceb));
    expect(scheme.surfaceVariant, isColor(0xff564050));
    expect(scheme.onSurfaceVariant, isColor(0xffdcbed2));
    expect(scheme.outline, isColor(0xffa4899b));
    expect(scheme.outlineVariant, isColor(0xff564050));
    expect(scheme.shadow, isColor(0xff000000));
    expect(scheme.scrim, isColor(0xff000000));
    expect(scheme.inverseSurface, isColor(0xfff4dceb));
    expect(scheme.inverseOnSurface, isColor(0xff3b2c37));
    expect(scheme.inversePrimary, isColor(0xffab00a2));
  });
}

File: cpp/scheme/scheme_test.cc
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cpp/scheme/scheme.h"

#include "testing/base/public/gunit.h"
#include "cpp/utils/utils.h"

namespace material_color_utilities {

namespace {

TEST(SchemeTest, SurfaceTones) {
  Argb color = 0xff0000ff;
  Scheme dark = MaterialDarkColorScheme(color);
  EXPECT_NEAR(LstarFromArgb(dark.surface), 10.0, 0.5);

  Scheme light = MaterialLightColorScheme(color);
  EXPECT_NEAR(LstarFromArgb(light.surface), 99.0, 0.5);
}

TEST(SchemeTest, BlueLightScheme) {
  Scheme scheme = MaterialLightColorScheme(0xff0000ff);
  EXPECT_EQ(HexFromArgb(scheme.primary), "ff343dff");
}

TEST(SchemeTest, BlueDarkScheme) {
  Scheme scheme = MaterialDarkColorScheme(0xff0000ff);
  EXPECT_EQ(HexFromArgb(scheme.primary), "ffbec2ff");
}

TEST(SchemeTest, ThirdPartyLightScheme) {
  Scheme scheme = MaterialLightColorScheme(0xff6750A4);
  EXPECT_EQ(HexFromArgb(scheme.primary), "ff6750a4");
  EXPECT_EQ(HexFromArgb(scheme.secondary), "ff625b71");
  EXPECT_EQ(HexFromArgb(scheme.tertiary), "ff7e5260");
  EXPECT_EQ(HexFromArgb(scheme.surface), "fffffbff");
  EXPECT_EQ(HexFromArgb(scheme.on_surface), "ff1c1b1e");
}

TEST(SchemeTest, ThirdPartyDarkScheme) {
  Scheme scheme = MaterialDarkColorScheme(0xff6750A4);
  EXPECT_EQ(HexFromArgb(scheme.primary), "ffcfbcff");
  EXPECT_EQ(HexFromArgb(scheme.secondary), "ffcbc2db");
  EXPECT_EQ(HexFromArgb(scheme.tertiary), "ffefb8c8");
  EXPECT_EQ(HexFromArgb(scheme.surface), "ff1c1b1e");
  EXPECT_EQ(HexFromArgb(scheme.on_surface), "ffe6e1e6");
}

TEST(SchemeTest, LightSchemeFromHighChromaColor) {
  Scheme scheme = MaterialLightColorScheme(0xfffa2bec);
  EXPECT_EQ(HexFromArgb(scheme.primary), "ffab00a2");
  EXPECT_EQ(HexFromArgb(scheme.on_primary), "ffffffff");
  EXPECT_EQ(HexFromArgb(scheme.primary_container), "ffffd7f3");
  EXPECT_EQ(HexFromArgb(scheme.on_primary_container), "ff390035");
  EXPECT_EQ(HexFromArgb(scheme.secondary), "ff6e5868");
  EXPECT_EQ(HexFromArgb(scheme.on_secondary), "ffffffff");
  EXPECT_EQ(HexFromArgb(scheme.secondary_container), "fff8daee");
  EXPECT_EQ(HexFromArgb(scheme.on_secondary_container), "ff271624");
  EXPECT_EQ(HexFromArgb(scheme.tertiary), "ff815343");
  EXPECT_EQ(HexFromArgb(scheme.on_tertiary), "ffffffff");
  EXPECT_EQ(HexFromArgb(scheme.tertiary_container), "ffffdbd0");
  EXPECT_EQ(HexFromArgb(scheme.on_tertiary_container), "ff321207");
  EXPECT_EQ(HexFromArgb(scheme.error), "ffba1a1a");
  EXPECT_EQ(HexFromArgb(scheme.on_error), "ffffffff");
  EXPECT_EQ(HexFromArgb(scheme.error_container), "ffffdad6");
  EXPECT_EQ(HexFromArgb(scheme.on_error_container), "ff410002");
  EXPECT_EQ(HexFromArgb(scheme.background), "fffffbff");
  EXPECT_EQ(HexFromArgb(scheme.on_background), "ff1f1a1d");
  EXPECT_EQ(HexFromArgb(scheme.surface), "fffffbff");
  EXPECT_EQ(HexFromArgb(scheme.on_surface), "ff1f1a1d");
  EXPECT_EQ(HexFromArgb(scheme.surface_variant), "ffeedee7");
  EXPECT_EQ(HexFromArgb(scheme.on_surface_variant), "ff4e444b");
  EXPECT_EQ(HexFromArgb(scheme.outline), "ff80747b");
  EXPECT_EQ(HexFromArgb(scheme.outline_variant), "ffd2c2cb");
  EXPECT_EQ(HexFromArgb(scheme.shadow), "ff000000");
  EXPECT_EQ(HexFromArgb(scheme.scrim), "ff000000");
  EXPECT_EQ(HexFromArgb(scheme.inverse_surface), "ff342f32");
  EXPECT_EQ(HexFromArgb(scheme.inverse_on_surface), "fff8eef2");
  EXPECT_EQ(HexFromArgb(scheme.inverse_primary), "ffffabee");
}

TEST(SchemeTest, DarkSchemeFromHighChromaColor) {
  Scheme scheme = MaterialDarkColorScheme(0xfffa2bec);
  EXPECT_EQ(HexFromArgb(scheme.primary), "ffffabee");
  EXPECT_EQ(HexFromArgb(scheme.on_primary), "ff5c0057");
  EXPECT_EQ(HexFromArgb(scheme.primary_container), "ff83007b");
  EXPECT_EQ(HexFromArgb(scheme.on_primary_container), "ffffd7f3");
  EXPECT_EQ(HexFromArgb(scheme.secondary), "ffdbbed1");
  EXPECT_EQ(HexFromArgb(scheme.on_secondary), "ff3e2a39");
  EXPECT_EQ(HexFromArgb(scheme.secondary_container), "ff554050");
  EXPECT_EQ(HexFromArgb(scheme.on_secondary_container), "fff8daee");
  EXPECT_EQ(HexFromArgb(scheme.tertiary), "fff5b9a5");
  EXPECT_EQ(HexFromArgb(scheme.on_tertiary), "ff4c2619");
  EXPECT_EQ(HexFromArgb(scheme.tertiary_container), "ff663c2d");
  EXPECT_EQ(HexFromArgb(scheme.on_tertiary_container), "ffffdbd0");
  EXPECT_EQ(HexFromArgb(scheme.error), "ffffb4ab");
  EXPECT_EQ(HexFromArgb(scheme.on_error), "ff690005");
  EXPECT_EQ(HexFromArgb(scheme.error_container), "ff93000a");
  EXPECT_EQ(HexFromArgb(scheme.on_error_container), "ffffb4ab");
  EXPECT_EQ(HexFromArgb(scheme.background), "ff1f1a1d");
  EXPECT_EQ(HexFromArgb(scheme.on_background), "ffeae0e4");
  EXPECT_EQ(HexFromArgb(scheme.surface), "ff1f1a1d");
  EXPECT_EQ(HexFromArgb(scheme.on_surface), "ffeae0e4");
  EXPECT_EQ(HexFromArgb(scheme.surface_variant), "ff4e444b");
  EXPECT_EQ(HexFromArgb(scheme.on_surface_variant), "ffd2c2cb");
  EXPECT_EQ(HexFromArgb(scheme.outline), "ff9a8d95");
  EXPECT_EQ(HexFromArgb(scheme.outline_variant), "ff4e444b");
  EXPECT_EQ(HexFromArgb(scheme.shadow), "ff000000");
  EXPECT_EQ(HexFromArgb(scheme.scrim), "ff000000");
  EXPECT_EQ(HexFromArgb(scheme.inverse_surface), "ffeae0e4");
  EXPECT_EQ(HexFromArgb(scheme.inverse_on_surface), "ff342f32");
  EXPECT_EQ(HexFromArgb(scheme.inverse_primary), "ffab00a2");
}

TEST(SchemeTest, LightContentSchemeFromHighChromaColor) {
  Scheme scheme = MaterialLightContentColorScheme(0xfffa2bec);
  EXPECT_EQ(HexFromArgb(scheme.primary), "ffab00a2");
  EXPECT_EQ(HexFromArgb(scheme.on_primary), "ffffffff");
  EXPECT_EQ(HexFromArgb(scheme.primary_container), "ffffd7f3");
  EXPECT_EQ(HexFromArgb(scheme.on_primary_container), "ff390035");
  EXPECT_EQ(HexFromArgb(scheme.secondary), "ff7f4e75");
  EXPECT_EQ(HexFromArgb(scheme.on_secondary), "ffffffff");
  EXPECT_EQ(HexFromArgb(scheme.secondary_container), "ffffd7f3");
  EXPECT_EQ(HexFromArgb(scheme.on_secondary_container), "ff330b2f");
  EXPECT_EQ(HexFromArgb(scheme.tertiary), "ff9c4323");
  EXPECT_EQ(HexFromArgb(scheme.on_tertiary), "ffffffff");
  EXPECT_EQ(HexFromArgb(scheme.tertiary_container), "ffffdbd0");
  EXPECT_EQ(HexFromArgb(scheme.on_tertiary_container), "ff390c00");
  EXPECT_EQ(HexFromArgb(scheme.error), "ffba1a1a");
  EXPECT_EQ(HexFromArgb(scheme.on_error), "ffffffff");
  EXPECT_EQ(HexFromArgb(scheme.error_container), "ffffdad6");
  EXPECT_EQ(HexFromArgb(scheme.on_error_container), "ff410002");
  EXPECT_EQ(HexFromArgb(scheme.background), "fffffbff");
  EXPECT_EQ(HexFromArgb(scheme.on_background), "ff1f1a1d");
  EXPECT_EQ(HexFromArgb(scheme.surface), "fffffbff");
  EXPECT_EQ(HexFromArgb(scheme.on_surface), "ff1f1a1d");
  EXPECT_EQ(HexFromArgb(scheme.surface_variant), "ffeedee7");
  EXPECT_EQ(HexFromArgb(scheme.on_surface_variant), "ff4e444b");
  EXPECT_EQ(HexFromArgb(scheme.outline), "ff80747b");
  EXPECT_EQ(HexFromArgb(scheme.outline_variant), "ffd2c2cb");
  EXPECT_EQ(HexFromArgb(scheme.shadow), "ff000000");
  EXPECT_EQ(HexFromArgb(scheme.scrim), "ff000000");
  EXPECT_EQ(HexFromArgb(scheme.inverse_surface), "ff342f32");
  EXPECT_EQ(HexFromArgb(scheme.inverse_on_surface), "fff8eef2");
  EXPECT_EQ(HexFromArgb(scheme.inverse_primary), "ffffabee");
}

TEST(SchemeTest, DarkContentSchemeFromHighChromaColor) {
  Scheme scheme = MaterialDarkContentColorScheme(0xfffa2bec);
  EXPECT_EQ(HexFromArgb(scheme.primary), "ffffabee");
  EXPECT_EQ(HexFromArgb(scheme.on_primary), "ff5c0057");
  EXPECT_EQ(HexFromArgb(scheme.primary_container), "ff83007b");
  EXPECT_EQ(HexFromArgb(scheme.on_primary_container), "ffffd7f3");
  EXPECT_EQ(HexFromArgb(scheme.secondary), "fff0b4e1");
  EXPECT_EQ(HexFromArgb(scheme.on_secondary), "ff4b2145");
  EXPECT_EQ(HexFromArgb(scheme.secondary_container), "ff64375c");
  EXPECT_EQ(HexFromArgb(scheme.on_secondary_container), "ffffd7f3");
  EXPECT_EQ(HexFromArgb(scheme.tertiary), "ffffb59c");
  EXPECT_EQ(HexFromArgb(scheme.on_tertiary), "ff5c1900");
  EXPECT_EQ(HexFromArgb(scheme.tertiary_container), "ff7d2c0d");
  EXPECT_EQ(HexFromArgb(scheme.on_tertiary_container), "ffffdbd0");
  EXPECT_EQ(HexFromArgb(scheme.error), "ffffb4ab");
  EXPECT_EQ(HexFromArgb(scheme.on_error), "ff690005");
  EXPECT_EQ(HexFromArgb(scheme.error_container), "ff93000a");
  EXPECT_EQ(HexFromArgb(scheme.on_error_container), "ffffb4ab");
  EXPECT_EQ(HexFromArgb(scheme.background), "ff1f1a1d");
  EXPECT_EQ(HexFromArgb(scheme.on_background), "ffeae0e4");
  EXPECT_EQ(HexFromArgb(scheme.surface), "ff1f1a1d");
  EXPECT_EQ(HexFromArgb(scheme.on_surface), "ffeae0e4");
  EXPECT_EQ(HexFromArgb(scheme.surface_variant), "ff4e444b");
  EXPECT_EQ(HexFromArgb(scheme.on_surface_variant), "ffd2c2cb");
  EXPECT_EQ(HexFromArgb(scheme.outline), "ff9a8d95");
  EXPECT_EQ(HexFromArgb(scheme.outline_variant), "ff4e444b");
  EXPECT_EQ(HexFromArgb(scheme.shadow), "ff000000");
  EXPECT_EQ(HexFromArgb(scheme.scrim), "ff000000");
  EXPECT_EQ(HexFromArgb(scheme.inverse_surface), "ffeae0e4");
  EXPECT_EQ(HexFromArgb(scheme.inverse_on_surface), "ff342f32");
  EXPECT_EQ(HexFromArgb(scheme.inverse_primary), "ffab00a2");
}

}  // namespace
}  // namespace material_color_utilities
